MSI-1486: Create new infrastructure for StockItem Configuration

seruymt opened the pull request:
https://github.com/magento-engcom/msi/issues/1486

seruymt committed:
MSI-1486: Initial commit

seruymt committed:
MSI-1486: create ui on stock edit page and source edit page

aleron75 commented on app/code/Magento/InventoryConfiguration/Model/ResourceModel/GetConfigurationValue.php:
```diff
@@ -0,0 +1,77 @@
+            ->where('config_option = ?', $configOption)
+            ->limit(1);
+
+        if (isset($stockId)) {
```
Hi @seruymt, I don't understand why, when `$stockId`, `$sourceCode`, and `$sku` are set you add the `IS NULL` clause.
I may be wrong but shouldn't it be the opposite? 
Or am I missing something? 
Thank you

seruymt replied:
Hi @aleron75, thanks for your comment. This is a way how we divide config option on scopes (Stock, StockItem, Source, SourceItem and Global). You can read more about it here - https://github.com/magento-engcom/msi/wiki/Inventory-configuration-Design-and-DB-structure

aleron75 replied:
Hi @seruymt I've read the document you pointed out. There, for example, I read:

>To retrieve Source level configuration options we need to execute next query:
>`select * from inventory_configuration where source_code = %source_code% and sku is NULL and stock_id is NULL`

as far as I understand by the example above, to retrieve a value at source level we only use the following match: `source_code = %source_code%`

so I still miss the point in committed code when, for example at line 61 of `GetConfigurationValue` class, if `$sourceCode` is set, we generate a `$select->where('source_code IS NULL');` statement instead of a `$select->where('source_code = ?', $sourceCode);`

Thanks

seruymt replied:
@aleron75 oh, it's my fault.. you're totally right.. (in previous version I've used "empty" instead of "isset"). I'll fix it, thanks:)

aleron75 commented on app/code/Magento/InventoryConfigurationAdminUi/etc/adminhtml/di.xml:
```diff
@@ -0,0 +1,15 @@
+-->
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
+    <type name="Magento\InventoryAdminUi\Ui\DataProvider\StockDataProvider">
+        <plugin name="inventory_stock_configuration" type="Magento\InventoryConfigurationAdminUi\Plugin\InventoryAdminUi\Ui\StockDataProvider\StockConfigurationPlugin" />
```
We should choose plugin names that convey the idea of what they do; in this case what about using `name="add_stock_configuration_data"`?

aleron75 commented on app/code/Magento/InventoryConfigurationAdminUi/etc/adminhtml/di.xml:
```diff
@@ -0,0 +1,15 @@
+        <plugin name="inventory_stock_configuration" type="Magento\InventoryConfigurationAdminUi\Plugin\InventoryAdminUi\Ui\StockDataProvider\StockConfigurationPlugin" />
+    </type>
+    <type name="Magento\InventoryAdminUi\Ui\DataProvider\SourceDataProvider">
+        <plugin name="inventory_source_configuration" type="Magento\InventoryConfigurationAdminUi\Plugin\InventoryAdminUi\Ui\SourceDataProvider\SourceConfigurationPlugin" />
```
We should choose plugin names that convey the idea of what they do; in this case what about using `name="add_source_configuration_data"`?

aleron75 commented on app/code/Magento/InventoryConfigurationAdminUi/etc/module.xml:
```diff
@@ -0,0 +1,10 @@
+ */
+-->
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
+    <module name="Magento_InventoryConfigurationAdminUi" setup_version="1.0.0"/>
```
We don't need to use the `setup_version` since we are using declarative schema definition

seruymt commented:
@aleron75 Also, this is just a prototype, we still work on it:) If you have any ideas how to improve interfaces or other parts of InventoryConfiguration feel free to share with us

seruymt committed:
MSI-1486: improvements

seruymt committed:
MSI-1486: InventoryConfiguration - new way (interfaces only)

seruymt committed:
MSI-1486: new interfaces

Commit:
MSI-1486: Create new infrastructure for StockItem Configuration

lorikrell commented:
I am working on documentation drafts for the new options, and have a question. 

The custom Source and Stock pages and Product Advanced Options have checkboxes for **Use Config Settings**. Is this the default config options set on Source and Stock? Or the default System Configs set in Stores > Configuration > Catalog > Inventory? Or was this just a naming issue for the checkbox, and it should be **Use system value**?

![image](https://user-images.githubusercontent.com/29388020/45106021-3f4b9d00-b0fb-11e8-92bd-e552c2a88906.png)

Commit:
MSI-1486: Create new infrastructure for StockItem Configuration

Commit:
MSI-1486: Create new infrastructure for StockItem Configuration

Commit:
MSI-1486: Create new infrastructure for StockItem Configuration

Commit:
Merge branch '2.3-develop' into MSI-1486

# Conflicts:
#	app/code/Magento/InventoryLowQuantityNotificationAdminUi/view/adminhtml/ui_component/product_form.xml
#	composer.lock

nmalevanec commented:
@lorikrell, There are several layers of configuration. 
Magento configuration consist of system values configuration(stored in system.xml and it's basically default magento configuration after install and cannot be changed from admin ui) and system configuration(stored in DB) and may be changed by admin user on magento configuration page.
So, if we check "use config settings" on Stock page configuration tab, it means Magento will fall back to system configuration to get the value. And if won't find value in DB("use system value" checked on configuration page for instance) will fall back to system value configuration, stored in different system.xml files.

So, the longest chain currently may be:
On Product Page "Advanced Inventory" panel for "Only X Left Threshold" field "use config settings" is checked.
Magento will fall back to corresponding stock configuration for given product.
If on Stock Page configuration tab for "Only X Left Threshold" field "use config settings" also is checked, magento will fall back to system configuration.
If on Configuration page for "Only X Left Threshold" field "use system value" is checked - value will be retrieved from system.xml file.

lorikrell commented:
Thank you for the info, @nmalevanec! That's great info!

I understand the fall back relationship of Product setting > Source settings / Stock settings > System settings. The naming of the option was different than the setting everywhere else. Just wanted to make sure it was intended. 

I'll add detailed info based on your feedback, and give a diagram to users for understanding it. And will update the docs with the checkbox name.

Commit:
MSI-1486: Create new infrastructure for StockItem Configuration

Commit:
MSI-1486: Create new infrastructure for StockItem Configuration

Commit:
MSI-1486: Create new infrastructure for StockItem Configuration

nmalevanec committed:
Merge branch '2.3-develop' into MSI-1486

nmalevanec committed:
MSI-1486: Create new infrastructure for StockItem Configuration

nmalevanec committed:
Merge remote-tracking branch 'origin/MSI-1486' into MSI-1486

nmalevanec committed:
MSI-1486: Create new infrastructure for StockItem Configuration

nmalevanec committed:
MSI-1486: Create new infrastructure for StockItem Configuration

nmalevanec committed:
MSI-1486: Create new infrastructure for StockItem Configuration

nmalevanec committed:
Merge branch '2.3-develop' into MSI-1486

Commit:
MSI-1486: Create new infrastructure for StockItem Configuration

nmalevanec committed:
MSI-1486: Create new infrastructure for StockItem Configuration

nmalevanec committed:
MSI-1486: Create new infrastructure for StockItem Configuration

nmalevanec committed:
MSI-1486: Create new infrastructure for StockItem Configuration

nmalevanec committed:
MSI-1486: Create new infrastructure for StockItem Configuration

nmalevanec commented:
Known issues:
1. Stock item in core magento has default values set on DB level. So, null value in model could be some default value for new stock item. Like "use_config_value_for_something", if not set - will be 1 for new stock item.(partially solved)
2. dev/tests/integration/testsuite/Magento/ConfigurableProduct/Model/Product/VariationHandlerTest::testGenerateSimpleProductsWithPartialData failing because test tries to generate simple product with qty = null and expecting it to be in stock. Which is not possible anymore because of fall back mechanism
https://github.com/magento-engcom/msi/blob/5877c063174e2d15b4514e332950b826d6c96feb/app/code/Magento/CatalogInventory/Model/StockStateProvider.php#L71-L81
$stockItem->getManageStock() initially not set in $stockItem model is falling back to system configuration now.
4. Hydration with non-existing product. dev/tests/integration/testsuite/Magento/Framework/Model/Entity/HydratorTest.php
Currently operations with stock item which has product id require this product to be in system.
4. Performance toolkit generation. Looks like Magento\Setup\Model\FixtureGenerator\ProductGeneratorTest expecting some generator for stock and source item configuration, caused by product creation is now involves stock and source configuration save.

maghamed commented on app/code/Magento/InventoryConfigurationApi/Api/GetInventoryConfigurationInterface.php:
```diff
@@ -0,0 +1,80 @@
+
+namespace Magento\InventoryConfigurationApi\Api;
+
+interface GetInventoryConfigurationInterface
```
This interface should look like 

```
interface GetInventoryConfigurationInterface
{
    /**
     * @return bool
     */
    public function isQtyDecimal($sku, $stockId): bool;

    /**
     * @return bool
     */
    public function isShowDefaultNotificationMessage($sku, $stockId): bool;

    /**
     * @return float
     */
    public function getMinQty($sku, $stockId): float;

    /**
     * @return float
     */
    public function getMinSaleQty($sku, $stockId): float;

    /**
     * @return float
     */
    public function getMaxSaleQty($sku, $stockId): float;

    /**
     * Retrieve backorders status
     *
     * @return int
     */
    public function getBackorders($sku, $stockId): int;

    /**
     * @return float
     */
    public function getNotifyStockQty($sku, $stockId): float;

    /**
     * Retrieve Quantity Increments data wrapper
     *
     * @return float
     */
    public function getQtyIncrements($sku, $stockId): float;

    /**
     * @return bool
     */
    public function isEnableQtyIncrements($sku, $stockId): bool;

    /**
     * @return bool
     */
    public function isManageStock($sku, $stockId): bool;

    /**
     * @return string
     */
    public function getLowStockDate($sku, $stockId): string;

    /**
     * @return bool
     */
    public function isDecimalDivided($sku, $stockId): bool;

    /**
     * @return float
     */
    public function getStockThresholdQty($sku, $stockId): float;
}

```

maghamed reviewed (changes requested):
Magento\InventorySales module should rely on 
`\Magento\InventoryConfigurationApi\Api\GetInventoryConfigurationInterface`

but not on 
- GetSourceConfigurationInterface
- GetStockConfigurationInterface

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration (temporary fix for checkin tests)

seruymt committed:
Merge branch '2.3-develop' into MSI-1486

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
Merge branch '2.3-develop' into MSI-1486

# Conflicts:
#	app/code/Magento/InventoryCatalogAdminUi/Model/CanManageSourceItemsBySku.php
#	app/code/Magento/InventoryLowQuantityNotificationAdminUi/view/adminhtml/ui_component/product_form.xml

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration

phoenix128 committed:
FIX error while trying to save a stock configuration

phoenix128 committed:
FIX error while opening the product's page

phoenix128 committed:
Merge branch '2.3-develop' into MSI-1486

phoenix128 committed:
FIX wrong usage of getTableName from connection instead of resource connection
NOTE: This bug was causing a side effect tests failure on fixtures

phoenix128 committed:
FIX products fixture with default inventory configuration and rollback for missing product removal

phoenix128 committed:
FIX nullable fields
All the fields should be nullable to fallback on default configuration

phoenix128 committed:
FIX missing fields in inventory configuration

phoenix128 committed:
FIX: Incremental configuration was not considered when validating

phoenix128 committed:
FIX failing tests

phoenix128 committed:
Removed conflicting product_form.xml from InventoryLowQuantityNotificationAdminUi

phoenix128 committed:
Removed stock items configuration from quantity increment tests to only rely on default stock configuration

maghamed commented:
I've run all test builds on Jenkins, accessible by this link - https://m2build-ur.devops.magento.com/job/All-User-Requested-Tests/5791/

phoenix128 committed:
FIX schema failure
This is a temporary fix that should be investigated in the Magento 2.3 core

phoenix128 committed:
Merge branch '2.3-develop' into MSI-1486

phoenix128 committed:
Wrong database isolation between tests or missing cleanups
Some of the tests were turning off database isolation for no reason and others were not cleaning modified database information

maghamed committed:
Merge pull request #1721 from magento-engcom/MSI-1486-tests-isolation

Wrong database isolation between tests or missing cleanups

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration (fix error with fixture generation test)

maghamed committed:
Merge pull request #1725 from magento-engcom/MSI-1486-1

MSI-1486: Create new infrastructure for StockItem Configuration (fix …

phoenix128 committed:
Merge branch '2.3-develop' into MSI-1486

phoenix128 committed:
Teardown for cleaning

phoenix128 committed:
Merge branch 'MSI-1486' into MSI-1486-tests-isolation

phoenix128 committed:
Test fixes

phoenix128 committed:
Removed unnecessary source settings and teardowns
The applied settings were equivalent to default settings, so they are useless.

phoenix128 committed:
Missing app isolation in testMultipleStores was causing a failure in testIsSalable

phoenix128 committed:
Removed isolation due to transaction issues
Isolation is not needed anymore in the modified tests and it is causing issues with DDL operations

maghamed committed:
Merge pull request #1727 from magento-engcom/MSI-1486-tests-isolation

Msi 1486 tests isolation

maghamed commented:
latest tests run on Jenkins - https://m2build-ur.devops.magento.com/job/All-User-Requested-Tests/6177/

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration (mftf test fix)

seruymt committed:
Merge branch '2.3-develop' into MSI-1486

seruymt committed:
MSI-1486: Create new infrastructure for StockItem Configuration (fix failed api-functional test)

maghamed commented:
link to the latest builds run https://m2build-ur.devops.magento.com/job/All-User-Requested-Tests/6217/

maghamed commented:
tests on 2.3-develop for comparison https://m2build-ur.devops.magento.com/job/All-User-Requested-Tests/6340/

aleron75 commented:
@magento-engcom-team give me test instance

ishakhsuvarov commented:
Unfortunately we have to close this PR now as it seems very outdated and in a state beyond repair. Another effort to implement the same feature is now happening in https://github.com/magento/inventory/pull/2297

Thanks
